signup and login

lei29 opened the pull request:
implemented signup and login, go into routes folder and nodemon index.js to test the backend, follow the error msg to define jwtprivatekey. config folder location would be changed after we rewrite the app.js in root folder.

Commit:
signup and login

implemented signup and login, go into routes folder and nodemon index.js to test the backend, follow the error msg to define jwtprivatekey.

smithpg commented on server/routes/login.js:
```diff
@@ -0,0 +1,41 @@
+    const { error } = validate(req.body);
+    if (error) return res.status(400).send(error.details[0].message);
+    
+    let user = await Chef.findOne({ email:req.body.email });
```
We can use the User model here instead of attempting both the Chef and Customer subtypes. 
```suggestion
    let user = await User.findOne({email: req.body.email});
```

lei29 replied:
Changed and tested, codde is pushed in new commit

smithpg commented on server/routes/login.js:
```diff
@@ -0,0 +1,41 @@
+            return res.status(400).send("Invalid email or password");
+        }
+    }
+    const isvalidpassword = await bcrypt.compare(req.body.password, user.password);
```
The User model has a method that can accomplish this in a slightly more concise way. It's just a wrapper around bcrypt.compare.

```suggestion
    const isvalidpassword = await user.comparePassword(req.body.password);
```

lei29 replied:
Changed and tested, codde is pushed in new commit

smithpg commented on server/routes/login.js:
```diff
@@ -0,0 +1,41 @@
+const {Chef,Customer} = require('../models/index')
```
```suggestion
const {User} = require('../models/index')
```

smithpg commented on server/routes/signup.js:
```diff
@@ -0,0 +1,58 @@
+});
+
+
+function validate(user){
```
Can we build this validation into the User schema?

lei29 replied:
Yes I could do that, but that may need more code written, using Joi will be much easier doing the validation.

smithpg replied:
Yeah, we can always refactor later on if we have extra time

smithpg commented on server/routes/signup.js:
```diff
@@ -0,0 +1,58 @@
+//store user in db depends on user type. return bad request or token and user.
+async function storeUser(usermodel,req,res){
+    
+    let chefEmail = await Chef.findOne({ email:req.body.email });
```
Mongo will throw an error if we violate the uniqueness rules for email or username. I'm pretty sure we can just attempt to create the user from req.body inside a try block and then catch the error if it turns out the email or username wasn't unique.

Commit:
changed requested change

change to User.findOne() , use try catch to avoid the uniqueness in signup process.
